File: programs/an-smart-contracts/src/mint.rs
use anchor_lang::prelude::*;
use anchor_spl::{
    associated_token::AssociatedToken,
    metadata::{
        create_master_edition_v3, create_metadata_accounts_v3, CreateMasterEditionV3,
        CreateMetadataAccountsV3, Metadata,
    },
    token::{self, Mint, Token, TokenAccount},
};
use mpl_token_metadata::types::DataV2;

pub fn mint_nft(ctx: Context<MintNft>, name: String, symbol: String) -> Result<()> {
    msg!("Minting token to token account...");
    msg!("Mint: {}", &ctx.accounts.mint.to_account_info().key());
    msg!(
        "Token Address: {}",
        &ctx.accounts.associated_token_account.key()
    );

    token::mint_to(
        CpiContext::new(
            ctx.accounts.token_program.to_account_info(),
            token::MintTo {
                mint: ctx.accounts.mint.to_account_info(),
                to: ctx.accounts.associated_token_account.to_account_info(),
                authority: ctx.accounts.signer.to_account_info(),
            },
        ),
        1,
    )?;

    let cpi_context = CpiContext::new(
        ctx.accounts.token_metadata_program.to_account_info(),
        CreateMetadataAccountsV3 {
            metadata: ctx.accounts.metadata_account.to_account_info(),
            mint: ctx.accounts.mint.to_account_info(),
            mint_authority: ctx.accounts.signer.to_account_info(),
            update_authority: ctx.accounts.signer.to_account_info(),
            payer: ctx.accounts.signer.to_account_info(),
            system_program: ctx.accounts.system_program.to_account_info(),
            rent: ctx.accounts.rent.to_account_info(),
        },
    );

    let data_v2 = DataV2 {
        name,
        symbol,
        uri: "https://alphaneural.io/".to_string(),
        seller_fee_basis_points: 0,
        creators: None,
        collection: None,
        uses: None,
    };

    create_metadata_accounts_v3(cpi_context, data_v2, false, true, None)?;

    let cpi_context = CpiContext::new(
        ctx.accounts.token_metadata_program.to_account_info(),
        CreateMasterEditionV3 {
            edition: ctx.accounts.master_edition_account.to_account_info(),
            mint: ctx.accounts.mint.to_account_info(),
            update_authority: ctx.accounts.signer.to_account_info(),
            payer: ctx.accounts.signer.to_account_info(),
            mint_authority: ctx.accounts.signer.to_account_info(),
            metadata: ctx.accounts.metadata_account.to_account_info(),
            token_program: ctx.accounts.token_program.to_account_info(),
            system_program: ctx.accounts.system_program.to_account_info(),
            rent: ctx.accounts.rent.to_account_info(),
        },
    );

    create_master_edition_v3(cpi_context, None)?;

    Ok(())
}

#[derive(Accounts)]
pub struct MintNft<'info> {
    /// CHECK: ok, we are passing in this account ourselves
    #[account(mut, signer)]
    pub signer: AccountInfo<'info>,
    #[account(
        init,
        payer = signer,
        mint::decimals = 0,
        mint::authority = signer.key(),
        mint::freeze_authority = signer.key(),
    )]
    pub mint: Account<'info, Mint>,
    #[account(
        init,
        payer = signer,
        associated_token::mint = mint,
        associated_token::authority = signer
    )]
    pub associated_token_account: Account<'info, TokenAccount>,
    /// CHECK: We're about to create this with Metaplex
    #[account(mut)]
    pub metadata_account: AccountInfo<'info>,
    /// CHECK: We're about to create this with Metaplex
    #[account(mut)]
    pub master_edition_account: AccountInfo<'info>,
    pub token_program: Program<'info, Token>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    pub token_metadata_program: Program<'info, Metadata>,
    pub system_program: Program<'info, System>,
    pub rent: Sysvar<'info, Rent>,
}

File: tests/mint.ts
import * as anchor from "@coral-xyz/anchor";
import { Program } from "@coral-xyz/anchor";
import { PublicKey, Transaction } from "@solana/web3.js";
import { expect } from "chai";
import {
  AccountLayout,
  createAssociatedTokenAccountInstruction,
  createTransferInstruction,
  getAssociatedTokenAddress,
  MintLayout,
  TOKEN_PROGRAM_ID,
} from "@solana/spl-token";
import {
  findMasterEditionPda,
  findMetadataPda,
  mplTokenMetadata,
  deserializeMetadata,
} from "@metaplex-foundation/mpl-token-metadata";
import { AnSmartContracts } from "../target/types/an_smart_contracts";
import { walletAdapterIdentity } from "@metaplex-foundation/umi-signer-wallet-adapters";
import { publicKey } from "@metaplex-foundation/umi";
import { createUmi } from "@metaplex-foundation/umi-bundle-defaults";

describe("fn mint", () => {
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);

  const signer = provider.wallet as anchor.Wallet;
  const program = anchor.workspace.AnSmartContracts as Program<AnSmartContracts>;

  const umi = createUmi("https://api.apr.dev").use(walletAdapterIdentity(signer)).use(mplTokenMetadata());

  it("should mint an NFT with the correct properties", async () => {
    const mint = anchor.web3.Keypair.generate();

    let metadataPda = findMetadataPda(umi, {
      mint: publicKey(mint.publicKey),
    })[0];

    let masterEditionPda = findMasterEditionPda(umi, {
      mint: publicKey(mint.publicKey),
    })[0];

    const associatedTokenAccount = await getAssociatedTokenAddress(mint.publicKey, signer.publicKey);

    const name = "Kobeni";
    const symbol = "kBN";

    await program.methods
      .mintNft(name, symbol)
      .accounts({
        signer: signer.publicKey,
        mint: mint.publicKey,
        associatedTokenAccount,
        metadataAccount: metadataPda,
        masterEditionAccount: masterEditionPda,
        // ** the following accounts are placed by anchor automatically
        // tokenMetadataProgram: MPL_TOKEN_METADATA_PROGRAM_ID,
        // tokenProgram: TOKEN_PROGRAM_ID,
        // associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
        // systemProgram: anchor.web3.SystemProgram.programId,
        // rent: anchor.web3.SYSVAR_RENT_PUBKEY,
        // alphaneuralProgram: program.programId,
      })
      .signers([mint])
      .rpc();

    // associated token account
    const rawTokenAccount = await provider.connection.getAccountInfo(associatedTokenAccount);
    const decodedTokenAccount = AccountLayout.decode(Buffer.from(rawTokenAccount.data));

    const rawMint = await provider.connection.getAccountInfo(mint.publicKey);
    const decodedMint = MintLayout.decode(Buffer.from(rawMint.data));

    expect(decodedTokenAccount.mint.toString()).equal(mint.publicKey.toString());
    expect(decodedTokenAccount.owner.toString()).equal(signer.publicKey.toString());
    expect(decodedTokenAccount.amount.toString()).equal("1"); // 1 nft
    expect(decodedTokenAccount.delegateOption).equal(0);
    expect(decodedTokenAccount.closeAuthorityOption).equal(0);
    expect(decodedTokenAccount.state).equal(1);

    // mint account
    expect(decodedMint.mintAuthority.toString()).equal(masterEditionPda.toString());
    expect(decodedMint.freezeAuthority.toString()).equal(masterEditionPda.toString());
    expect(decodedMint.supply.toString()).equal("1");
    expect(decodedMint.isInitialized).equal(true);

    // metadata account
    const rawMetadata = await provider.connection.getAccountInfo(new PublicKey(metadataPda.toString()));
    const decodedMetadata = deserializeMetadata(rawMetadata as any);
    expect(decodedMetadata.mint.toString()).equal(mint.publicKey.toString());
    expect(decodedMetadata.symbol.length).greaterThan(0);

    expect((decodedMetadata.creators as any).value[0].address).equal(signer.publicKey.toString());
    expect((decodedMetadata.creators as any).value[0].share).equal(100);
  });

  it("should mint an NFT and transfer it successfully", async () => {
    const mint = anchor.web3.Keypair.generate();
    const receiver = anchor.web3.Keypair.generate();

    let metadataPda = findMetadataPda(umi, {
      mint: publicKey(mint.publicKey),
    })[0];

    let masterEditionPda = findMasterEditionPda(umi, {
      mint: publicKey(mint.publicKey),
    })[0];

    const associatedTokenAccount = await getAssociatedTokenAddress(mint.publicKey, signer.publicKey);

    const receiverTokenAccount = await getAssociatedTokenAddress(mint.publicKey, receiver.publicKey);

    const name = "Kobeni";
    const symbol = "kBN";

    await program.methods
      .mintNft(name, symbol)
      .accounts({
        signer: signer.publicKey,
        mint: mint.publicKey,
        associatedTokenAccount,
        metadataAccount: metadataPda,
        masterEditionAccount: masterEditionPda,
      })
      .signers([mint])
      .rpc();

    // Ensure the associated token account for the receiver is created
    const ataTx = new Transaction().add(
      createAssociatedTokenAccountInstruction(signer.publicKey, receiverTokenAccount, receiver.publicKey, mint.publicKey)
    );
    await provider.sendAndConfirm(ataTx, [signer.payer]);

    // Transfer the NFT to the receiver
    const transferTx = new Transaction().add(
      createTransferInstruction(associatedTokenAccount, receiverTokenAccount, signer.publicKey, 1, [], TOKEN_PROGRAM_ID)
    );
    await provider.sendAndConfirm(transferTx, [signer.payer]);

    // Verify the receiver's token account
    const rawReceiverTokenAccount = await provider.connection.getAccountInfo(receiverTokenAccount);
    const decodedReceiverTokenAccount = AccountLayout.decode(Buffer.from(rawReceiverTokenAccount.data));

    expect(decodedReceiverTokenAccount.mint.toString()).equal(mint.publicKey.toString());
    expect(decodedReceiverTokenAccount.owner.toString()).equal(receiver.publicKey.toString());
    expect(decodedReceiverTokenAccount.amount.toString()).equal("1"); // 1 NFT

    // Verify the original token account has no NFT
    const rawTokenAccount = await provider.connection.getAccountInfo(associatedTokenAccount);
    const decodedTokenAccount = AccountLayout.decode(Buffer.from(rawTokenAccount.data));

    expect(decodedTokenAccount.amount.toString()).equal("0"); // No NFT

    // Verify creator is still original signer
    const rawMetadata = await provider.connection.getAccountInfo(new PublicKey(metadataPda.toString()));
    const decodedMetadata = deserializeMetadata(rawMetadata as any);
    expect((decodedMetadata.creators as any).value[0].address).equal(signer.publicKey.toString());
  });
});
